### PR TITLE
Adjustable stack size for EVM

### PR DESCRIPTION
**Pull request by NikVolf**
#2482

### CODE REVIEW

File: ethcore/src/executive.rs [hunk 1]
```diff
@@ -25,10 +25,10 @@ use trace::{FlatTrace, Tracer, NoopTracer, ExecutiveTracer, VMTrace, VMTracer, E
 use crossbeam;
 pub use types::executed::{Executed, ExecutionResult};
 
-/// Max depth to avoid stack overflow (when it's reached we start a new thread with VM)
+/// Roughly estimate what stack size each level of evm depth will use
 /// TODO [todr] We probably need some more sophisticated calculations here (limit on my machine 132)
 /// Maybe something like here: `https://github.com/ethereum/libethereum/blob/4db169b8504f2b87f7d5a481819cfb959fc65f6c/libethereum/ExtVM.cpp`
-const MAX_VM_DEPTH_FOR_THREAD: usize = 64;
+const STACK_SIZE_PER_DEPTH: usize = 256*1024;
```
Comment (arkpar) [inline ethcore/src/executive.rs]: 256kb per recursion frame? That's too much. Before the change it was 512k / 64 = 8kb
512k is the minimum stack size that we supported

File: util/io/src/worker.rs [hunk 2]
```diff
@@ -22,9 +22,19 @@ use crossbeam::sync::chase_lev;
...
 
 use std::sync::{Condvar as SCondvar, Mutex as SMutex};
 
+const STACK_SIZE: usize = 16*1024*1024;
+
+thread_local! {
+	/// Stack size
+	/// Should be modified if it is changed in Rust since it is no way
+	/// to know or get it
+	pub static LOCAL_STACK_SIZE: Cell<usize> = Cell::new(::std::env::var("RUST_MIN_STACK").ok().and_then(|s| s.parse().ok()).unwrap_or(2 * 1024 * 1024));
```
Comment (NikVolf) [inline util/io/src/worker.rs]: 512k must be default for the main thread only

this expression exactly the same value which rust uses for spawned threads:
https://github.com/rust-lang/rust/blob/master/src/libstd/sys/common/util.rs#L18

File: ethcore/src/executive.rs (hunk 1 above)
Comment (NikVolf) [inline ethcore/src/executive.rs]: before change it was 2MB/64
i just used 16MB/64 instead, my bad

File: util/io/src/worker.rs [hunk 3]
```diff
@@ -22,9 +22,19 @@ use crossbeam::sync::chase_lev;
 use service::{HandlerId, IoChannel, IoContext};
 use IoHandler;
 use panics::*;
+use std::cell::Cell;
 
 use std::sync::{Condvar as SCondvar, Mutex as SMutex};
 
+const STACK_SIZE: usize = 16*1024*1024;
```
Comment (NikVolf) [inline util/io/src/worker.rs]: this might be so, but previously we used 32kb
see above

**Comment by NikVolf** [comment]
well 16kb was not enough
trying with 24kb

**Pull request merged**